various: Don't allow creation of invalid UTF8 strings or identifiers

jepler opened the pull request:
### Summary

Fuzz testing found that it was possible to create invalid UTF-8 strings when the program input was not UTF-8. This could occur because a disk file was not UTF-8, or because a byte string passed to eval()/exec() was not UTF-8.

Besides leading to the problems that the introduction of `utf8_check` was intended to fix (#9044), the fuzzer found an actual crash when the first byte was `\xff` and the string was used as an exception argument (#17855).

I also noticed that the check could be generalized a little to avoid constructing non-UTF-8 identifiers, which could also lead to problems.

I re-organized the code to pay for the size cost of the new check in the lexer.

### Testing

I added a new test, using `eval()` and `exec()` of byte strings, to ensure that these cases are caught by the lexer.

### Trade-offs and Alternatives

Could check that the whole code buffer is UTF-8 instead.

jepler committed:
py: Reduce code size from utf8_check.

All sites immediately threw a UnicodeError, so roll that into
the new function utf8_require.

unicode.c was designed not to require runtime.h, so move the
checking function into objstr.c.

Reduce the number of #if sites by making a do-nothing variant
that is used instead when !STR_UNICODE or !STR_UNICODE_CHECK.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
parse: Don't allow creation of invalid UTF8 strings or identifiers.

.. even when compiling non UTF-8 files or byte strings.

Closes: #17855
Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
tests: Add test of invalid unicode strings.

.. from non UTF-8 inputs. In this case, MicroPython raises
UnicodeError while CPython uses SyntaxError. By catching either
exception, the test does not require an .exp file.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler commented on tests/unicode/unicode_parser.py.exp:
```diff
@@ -0,0 +1,5 @@
+UnicodeError
```
Updated.

dpgeorge replied:
Would it be better to make MicroPython raise SyntaxError, to match CPython?

I don't know if that really matters though, this is a pretty rare case, and saving code size (reusing the same UnicodeError) is probably more important.

OTOH, if you had some code like this:
```py
try:
    exec(some_code)
except SyntaxError:
    handle_syntax_error()
```
you might be surprised that the UnicodeError is raised and escapes the error handling.

dpgeorge commented on py/parse.c:
```diff
@@ -598,6 +598,9 @@ static mp_parse_node_t make_node_const_object_optimised(parser_t *parser, size_t
     mp_parse_node_t pn;
     mp_lexer_t *lex = parser->lexer;
+    if (lex->tok_kind == MP_TOKEN_NAME || lex->tok_kind == MP_TOKEN_STRING) {
+        mp_utf8_require((byte *)lex->vstr.buf, lex->vstr.len);
```
This is going to impact performance for compilation of all code.  I guess there's not really any way around that if we want to validate the utf8 properly.

jepler replied:
Agreed on both counts. If the overhead is intolerable for some use case, it can be compile-time disabled just like the existing uses of mp_utf8_require (gated by MICROPY_PY_BUILTINS_STR_UNICODE_CHECK, on by default when unicode is on).

(It's been rattling around in my head to have a build flag for "remove all checks a good static checker would diagnose", e.g., code that passes `mypy --strict`, is checked for valid utf-8, etc. but I think there would still end up being a lot of judgement calls and also you can't quite type check micropython/circuitpython code because of lack of exactly matching pyi stubs…)

jepler commented:
I changed things a bit so SyntaxError can be raised in this case. I agree it seems preferable if the cost isn't high.

jepler committed:
objstr: Ensure utf8 in mp_obj_new_str_type_from_vstr.

This catches for instance the cases I found in #13084.
It does not bring the behavior in line with standard Python,
but it does throw errors in the case where a string object
would be created with invalid UTF-8 content.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
cpydiff: Document difference in {:c}, %c formatting.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler commented:
maybe I should back the string formatting stuff out again, it's not as clean a change as I thought. What do you think @dlech ?

dlech commented:
We could add a `bool` parameter to `mp_obj_new_str_type_from_vstr()` to optionally enable the check, but of course that comes as some small cost in code size and a few extra instructions executed.

jepler commented:
Closing this up, as it stands this is not a cleanup, it's just making things uglier. #17855 and (new) #18609 track the two issues of creating non UTF-8 string literals & identifiers.
